x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/optimizer/PhysicalPlanOptimizerTests.java

```diff
@@ -4023,6 +4024,25 @@ public void testRejectRemoteEnrichAfterCoordinatorEnrich() {
         );
     }
 
+    public void testNoStackOverflow() {
+        String query = randomBoolean() ? "row a = 1" : "row a = 1 | eval b = a";
+        for (int i = 0; i < 1900; i++) {
+            query += "::long::int";
+        }
+        for (int i = 0; i < 50; i++) {
+            for (int j = 0; j < randomInt(20); j++) {
+                query += "::long::int";
+            }
+
+            try {
+                physicalPlan(query);
+            } catch (ParsingException | org.elasticsearch.xpack.esql.parser.ParsingException e) {
```

Comment:
It's completely non-deterministic unfortunately. My feeling is that it also depends on how the JVM JIT inlines some methods, because based on how hot the env is and now many times I run the same test I see different a behavior (ie. sometimes it fails at a certain level, sometimes it needs much deeper expressions, sometimes it never fails)